crates/bws/src/command/run.rs

```diff
@@ -0,0 +1,135 @@
+use color_eyre::eyre::Result;
+use uuid::Uuid;
+use which::which;
+
+use crate::{
+    util::{is_valid_posix_name, uuid_to_posix},
+    Cli, ACCESS_TOKEN_KEY_VAR_NAME,
+};
+
+pub(crate) async fn run(
+    client: Client,
+    organization_id: Uuid,
+    project_id: Option<Uuid>,
+    uuids_as_keynames: bool,
+    no_inherit_env: bool,
+    shell: Option<String>,
+    command: Vec<String>,
+) -> Result<()> {
+    let shell = match std::env::consts::OS {
+        "windows" => shell.unwrap_or_else(|| "powershell".to_string()),
+        _ => shell.unwrap_or_else(|| "sh".to_string()),
+    };
+
+    if which(&shell).is_err() {
+        eprintln!("Error: shell '{}' not found", shell);
+        std::process::exit(1);
+    }
+
+    let user_command = if command.is_empty() {
+        if atty::is(Stream::Stdin) {
+            eprintln!("{}", Cli::command().render_help().ansi());
+            std::process::exit(1);
+        }
+
+        let mut buffer = String::new();
+        std::io::stdin().read_to_string(&mut buffer)?;
+        buffer
+    } else {
+        command.join(" ")
+    };
+
+    if user_command.is_empty() {
+        let mut cmd = Cli::command();
+        eprintln!("{}", cmd.render_help().ansi());
+        std::process::exit(1);
+    }
+
+    let res = if let Some(project_id) = project_id {
+        client
+            .secrets()
+            .list_by_project(&SecretIdentifiersByProjectRequest { project_id })
+            .await?
+    } else {
+        client
+            .secrets()
+            .list(&SecretIdentifiersRequest { organization_id })
+            .await?
+    };
+
+    let secret_ids = res.data.into_iter().map(|e| e.id).collect();
+    let secrets = client
+        .secrets()
+        .get_by_ids(SecretsGetRequest { ids: secret_ids })
+        .await?
+        .data;
+
+    let mut environment = std::collections::HashMap::new();
+
+    secrets.into_iter().for_each(|s| {
+        let key = if uuids_as_keynames {
+            uuid_to_posix(&s.id)
+        } else {
+            s.key
+        };
+
+        if !is_valid_posix_name(&key) {
+            eprintln!(
+                "Warning: secret '{}' does not have a POSIX-compliant name",
+                key
+            );
+        }
+
+        match environment.contains_key(&key) {
+            true => {
+                eprintln!(
+                    "Error: multiple secrets with name '{}' found. Use --uuids-as-keynames or use unique names for secrets.",
+                    key
+                );
+                std::process::exit(1);
+            }
+            false => {
+                environment.insert(key, s.value);
+            }
+        }
+    });
+
+    let mut command = process::Command::new(shell);
+    command
+        .arg("-c")
+        .arg(&user_command)
+        .stdout(process::Stdio::inherit())
+        .stderr(process::Stdio::inherit());
+
+    if no_inherit_env {
+        let path = std::env::var("PATH").unwrap_or_else(|_| "/bin:/usr/bin".to_string());
```

Comment:
**question:** Do we need to handle windows separately?

Reply:
Fixed in 23ab8f4b8551e53fb46a2e15b7a38e5a0dfdee95.